src/tls.py

```diff
@@ -0,0 +1,352 @@
+            )
+
+        # generate unit private key if not already created by action
+        if not self.keystore_password:
+            self.cluster.data[self.charm.unit].update({"keystore-password": generate_password()})
+
+        self._request_certificate()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handler for `certificates_available` event after provider updates signed certs."""
+        if not self.cluster:
+            event.defer()
+            return
+
+        # avoid setting tls files and restarting
+        if event.certificate_signing_request != self.csr:
+            logger.error("Can't use certificate, found unknown CSR")
+            return
+
+        # if certificate already exists, this event must be new, flag manual restart
+        if self.certificate:
+            self.cluster.data[self.charm.unit].update({"manual-restart": "true"})
+
+        self.cluster.data[self.charm.unit].update(
+            {"certificate": event.certificate, "ca": event.ca}
+        )
+
+        self.set_server_key()
+        self.set_ca()
+        self.set_certificate()
+        self.set_truststore()
+        self.set_p12_keystore()
+
+        self.charm.on[self.charm.restart.name].acquire_lock.emit()
+
+    def _on_certificates_broken(self, _) -> None:
+        """Handler for `certificates_relation_broken` event."""
+        self.cluster.data[self.charm.unit].update({"csr": "", "certificate": "", "ca": ""})
+
+        # remove all existing keystores from the unit so we don't preserve certs
+        self.remove_stores()
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # if this event fired, trigger `upgrade` from ssl -> non-ssl
+        # ideally trigger this before any other `certificates_*` step
+        self.cluster.data[self.charm.app].update({"tls": "", "upgrading": "started"})
+
+    def _on_certificate_expiring(self, _) -> None:
+        """Handler for `certificate_expiring` event."""
+        if not self.private_key or not self.csr:
+            logger.error("Missing unit private key and/or old csr")
+            return
+        new_csr = generate_csr(
+            private_key=self.private_key.encode("utf-8"),
+            subject=os.uname()[1],
+            sans=self._get_sans(),
+        )
+
+        self.certificates.request_certificate_renewal(
+            old_certificate_signing_request=self.csr.encode("utf-8"),
+            new_certificate_signing_request=new_csr,
+        )
+
+        self.cluster.data[self.charm.unit].update({"csr": new_csr.decode("utf-8").strip()})
+
+    def _set_tls_private_key(self, event: ActionEvent) -> None:
+        """Handler for `set_tls_private_key` action."""
+        private_key = self._parse_tls_file(event.params.get("internal-key", None))
+        self.cluster.data[self.charm.unit].update({"private-key": private_key})
+
+        self._on_certificate_expiring(event)
```

Comment:
Should this be changed to `self._request_certificate()? It seems that it may not succeed on L231 because of a missing old CSR.

Reply:
You're right, very good catch! However I found that requesting a new certificate failed the following check:

```python
        # avoid setting tls files and restarting
        if event.certificate_signing_request != self.csr:
            logger.error("Can't use certificate, found unknown CSR")
            return
```

I might look into it further.